### ISSUE TITLE
Heap is not a constructor

### ISSUE THREAD

**Issue by agaurav** [cytoscape/cytoscape.js#1952]
**Issue type**

Bug report

**Environment info**

- Cytoscape.js version : 3.2.1
- Browser/Node.js & version : 8.4.0
- webpack: 3.5.5

**Current (buggy) behaviour**

error: `Heap is not a constructor`
probably Webpack can't load the `heap` package. 

Relevant `tsconfig.json`

``` json
"compilerOptions": {
        "noUnusedParameters": true,
        "noUnusedLocals": true,
        "skipLibCheck": true,
        "noEmit": true,
        "target": "es2015",
        "module": "commonjs",
        "noImplicitAny": false,
        "sourceMap": true,
        "inlineSourceMap": false,
        "inlineSources": false,
        "preserveConstEnums": true,
        "moduleResolution": "node",
        "experimentalDecorators": true,
        "jsx": "preserve",
        "lib": [
            "dom",
            "es5",
            "es6",
            "es7",
            "es2017.object"
        ],
        "allowSyntheticDefaultImports": true
    },

```

Works fine with `v3.1.5`. Probably related to https://github.com/cytoscape/cytoscape.js/issues/1858

**Comment by maxkfranz** [comment]
It looks like it's a problem on your end.  Node works fine -- so browserify and webpack should work fine as well.

```
$ mkdir foobar
$ cd foobar
$ npm init -y
Wrote to /private/tmp/foobar/package.json:

{
  "name": "foobar",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}


$ npm i -S cytoscape
foobar@1.0.0 /private/tmp/foobar
└─┬ cytoscape@3.2.1
  ├── heap@0.2.6
  └── lodash.debounce@4.0.8

npm WARN foobar@1.0.0 No description
npm WARN foobar@1.0.0 No repository field.
$ node
> cytoscape = require('cytoscape')
{ [Function: cytoscape]
  use: [Function],
  version: '3.2.1',
  Stylesheet: [Function: Stylesheet],
  stylesheet: [Function: Stylesheet] }
> cy = cytoscape()
Core {
  _private:
   { container: undefined,
     ready: true,
     options: { layout: [Object], renderer: [Object] },
     elements: Collection { length: 0, _private: [Object] },
     listeners: [],
     aniEles: Collection { length: 0, _private: [Object] },
     scratch: {},
     layout: null,
     renderer: { options: [Object], notifications: 1 },
     destroyed: false,
     notificationsEnabled: true,
     minZoom: 1e-50,
     maxZoom: 1e+50,
     zoomingEnabled: true,
     userZoomingEnabled: true,
     panningEnabled: true,
     userPanningEnabled: true,
     boxSelectionEnabled: true,
     autolock: false,
     autoungrabify: false,
     autounselectify: false,
     styleEnabled: false,
     zoom: 1,
     pan: { x: 0, y: 0 },
     animation: { current: [], queue: [] },
     hasCompoundNodes: false,
     emitter:
      Emitter {
        qualifierCompare: [Function: qualifierCompare],
        eventMatches: [Function: eventMatches],
        eventFields: [Function: eventFields],
        callbackContext: [Function: callbackContext],
        beforeEmit: [Function: beforeEmit],
        afterEmit: [Function: afterEmit],
        bubble: [Function: bubble],
        parent: [Function: parent],
        context: [Circular],
        listeners: [] },
     selectionType: 'single',
     batchCount: 0,
     batchingNotify: false,
     batchingStyle: false,
     batchStyleEles: Collection { length: 0, _private: [Object] },
     batchNotifyEles: Collection { length: 0, _private: [Object] },
     batchNotifyTypes: [ ids: {} ],
     animationsRunning: true } }
> cy.add({ data: { id: 'a' } })
Collection {
  '0':
   Collection {
     '0': [Circular],
     length: 1,
     _private:
      { cy: [Object],
        single: true,
        data: [Object],
        position: [Object],
        autoWidth: undefined,
        autoHeight: undefined,
        autoPadding: undefined,
        compoundBoundsClean: false,
        listeners: [],
        group: 'nodes',
        style: {},
        rstyle: {},
        styleCxts: [],
        removed: false,
        selected: false,
        selectable: true,
        locked: false,
        grabbed: false,
        grabbable: true,
        active: false,
        classes: Set {},
        animation: [Object],
        rscratch: {},
        scratch: {},
        edges: [],
        children: [],
        parent: null,
        traversalCache: null,
        emitter: [Object],
        map: [Object] } },
  length: 1,
  _private:
   { cy: Core { _private: [Object] },
     map: Map { 'a' => [Object] } } }
> cy.add({ data: { id: 'b' } })
Collection {
  '0':
   Collection {
     '0': [Circular],
     length: 1,
     _private:
      { cy: [Object],
        single: true,
        data: [Object],
        position: [Object],
        autoWidth: undefined,
        autoHeight: undefined,
        autoPadding: undefined,
        compoundBoundsClean: false,
        listeners: [],
        group: 'nodes',
        style: {},
        rstyle: {},
        styleCxts: [],
        removed: false,
        selected: false,
        selectable: true,
        locked: false,
        grabbed: false,
        grabbable: true,
        active: false,
        classes: Set {},
        animation: [Object],
        rscratch: {},
        scratch: {},
        edges: [],
        children: [],
        parent: null,
        traversalCache: null,
        emitter: [Object],
        map: [Object] } },
  length: 1,
  _private:
   { cy: Core { _private: [Object] },
     map: Map { 'b' => [Object] } } }
> cy.add({ data: { id: 'ab', source: 'a', target: 'b' } })
Collection {
  '0':
   Collection {
     '0': [Circular],
     length: 1,
     _private:
      { cy: [Object],
        single: true,
        data: [Object],
        position: {},
        autoWidth: undefined,
        autoHeight: undefined,
        autoPadding: undefined,
        compoundBoundsClean: false,
        listeners: [],
        group: 'edges',
        style: {},
        rstyle: {},
        styleCxts: [],
        removed: false,
        selected: false,
        selectable: true,
        locked: false,
        grabbed: false,
        grabbable: true,
        active: false,
        classes: Set {},
        animation: [Object],
        rscratch: {},
        scratch: {},
        edges: [],
        children: [],
        parent: null,
        traversalCache: [Object],
        emitter: [Object],
        source: [Object],
        target: [Object],
        map: [Object] } },
  length: 1,
  _private:
   { cy: Core { _private: [Object] },
     map: Map { 'ab' => [Object] } } }
> cy.elements().dijkstra({ root: cy.nodes().first() })
{ distanceTo: [Function: distanceTo],
  pathTo: [Function: pathTo] }
>
```

`eles.dijkstra()` wouldn't work without the heap.  Something must be wrong with your TS config.

**Comment by agaurav** [comment]
The same configuration are working for all previous versions(`3.0.x`,`3.1.x`) so I thought the cause might be the changes in `v3.2.0`. I will look into the commits see if it can be narrowed down to the edit.

**Comment by maxkfranz** [comment]
Your config probably doesn't support transitive dependencies.

**Comment by opavader** [comment]
transitive dependencies are working for other packages. If it helps using `dist/cytoscape.js` instead of `dist/cytoscape.cjs.js` solves the issue.

**Comment by maxkfranz** [comment]
`dist/cytoscape.js` has bundled dependencies for the download ZIP.  It's only for people who use script tags on simple pages.

Cytoscape 3.2 is confirmed working with Node, Browserify, and Webpack.  You must have something in your own Typescript options that is breaking your build.  Try starting from a simpler, default config and work up to see where you have a mistake in your config.

If you can come up with a minimal example that shows that the problem is with the build of Cytoscape by Webpack, then we'll address it.  But I strongly suspect you've just made a mistake in your build config with TS, so I'm closing the issue for now.

**Comment by jaydenlin** [comment]
I have the same issue.
I used `dist/cytoscape.js` instead of `dist/cytoscape.cjs.js` to solve the issue.

**Comment by maxkfranz** [comment]
Cytoscape is not a Typescript library.  It's a JS library.  You have to configure your build properly to accommodate for that.

You have a problem with transitive dependencies.  Check your typings.  Check that you have typings for the npm dependencies of `cytoscape@3.2`.

I recommend you read the docs carefully for your tooling:

- [Typescript](https://www.typescriptlang.org/docs/home.html)
- [browserify tsify](https://github.com/TypeStrong/tsify)
- [webpack ts-loader](https://webpack.js.org/guides/typescript/)

It works for me with just a single line: 

```
browserify main.ts -p [ tsify --allowJs ] > bundle.js
```

For reference:

```js
// main.ts

import * as cytoscape from 'cytoscape';

let cy = cytoscape();

cy.add([
  { data: { id: 'foo' } },
  { data: { id: 'bar' } },
  { data: { source: 'foo', target: 'bar' } }
]);

console.log( 'num eles', cy.elements().size() );

console.log( 'dijkstra return obj', cy.elements().dijkstra({ root: '#foo' }) );
```

And if we run the bundle, it works just fine:

```
$ node bundle.js
num eles 3
dijkstra return obj { distanceTo: [Function: distanceTo],
  pathTo: [Function: pathTo] }
```

**Comment by agaurav** [comment]
I very much appreciate you putting extra effort to have this solved but for now I have made peace with using the bundled version :)

While it sure seems to be an issue with my `tsconfig` (or may be `webpack` config), but I am unable to identify it.
 I tried various combination for compiler option `target`,`module` & `allowSyntheticDefaultImports`, `allowJs` & `isolatedModules` but none seems to be able to work. The imported `Heap` variable seems to be an empty object,when editing the dist file as 
``` javascript
var Heap = __webpack_require__(8);
console.log(Heap);
```

The peer dependencies are working for all other of packages (almost all of them having react,d3 & lodash as one). I also have both `"@types/heap": "^0.2.28"` & `"heap": "^0.2.6"` installed.
Thanks to bundles version it's not an issues for me as `heap` is quite a small lib.

Again thanks for your help.

For anyone else facing same issue with webpack & typescript, a workaround it to have the following in your webpack config.
``` javascript
alias: { 
    cytoscape: path.join(process.cwd(), 'node_modules/cytoscape/dist/cytoscape.js'),
},
```
